File: LinkedList/Cpp/Problem8.cpp
/*
Insert at the end of Doubly LL.
*/

#include <iostream>
using namespace std;

struct Node
{
  int data;
  Node *next;
  Node *prev;
  Node(int x)
  {
    data = x;
    next = NULL;
    prev = NULL;
  }
};

Node *insertEnd(Node *head, int x)
{
  Node *temp = new Node(x);
  if (head == NULL)
  {
    return temp;
  }
  Node *curr = head;
  while (curr->next != NULL)
  {
    curr = curr->next;
  }
  curr->next = temp;
  temp->prev = curr;
  return head;
}

void printList(Node *node)
{
  while (node != NULL)
  {
    cout << node->data << " ";
    node = node->next;
  }
  cout << endl;
}

int main()
{
  Node *head = NULL;
  head = insertEnd(head, 10);
  head = insertEnd(head, 20);
  head = insertEnd(head, 30);
  head = insertEnd(head, 40);
  head = insertEnd(head, 50);
  head = insertEnd(head, 60);
  printList(head);
  return 0;
}

File: LinkedList/Java/Problem8.java
/*
Insert at the end of Doubly LL.
*/

import java.io.*;

class DLL {
  int data;
  DLL next;
  DLL prev;

  DLL(int x) {
    data = x;
    next = null;
    prev = null;
  }
}

class GFG {
  public static DLL insertEnd(DLL head, int x) {
    DLL temp = new DLL(x);
    if (head == null) {
      return temp;
    }
    DLL curr = head;
    while (curr.next != null) {
      curr = curr.next;
    }
    curr.next = temp;
    temp.prev = curr;
    return head;
  }

  public static void printList(DLL node) {
    while (node != null) {
      System.out.print(node.data + " ");
      node = node.next;
    }
    System.out.println("");
  }

  public static void main(String[] args) {
    DLL head = null;
    head = insertEnd(head, 10);
    head = insertEnd(head, 20);
    head = insertEnd(head, 30);
    head = insertEnd(head, 40);
    head = insertEnd(head, 50);
    head = insertEnd(head, 60);
    head = insertEnd(head, 70);
    head = insertEnd(head, 80);
    head = insertEnd(head, 90);
    printList(head);
  }
}
